File: scala/61.scala
object Solution {
    def rotateRight(head: ListNode, k: Int): ListNode = {
        if (head == null) null 
        else {
          var ptr = head 
          var length = 1
          var tail = ptr
          while (ptr.next != null) {
            length += 1
            ptr = ptr.next
          }
          tail = ptr
          var moves = (length - k % length)%length
          ptr = head
          while(moves >0){
            val next = ptr.next
            ptr.next = null
            tail.next = ptr
            tail = tail.next
            ptr = next
            moves -= 1
          }
          ptr
        }
    }
}

File: leetcodeJava/61.java
package leetcodeJava;

class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        var ptr = head;
        var tail = head;
        var l = 0;
        while (ptr != null) {
            l++;
            tail = ptr;
            ptr = ptr.next;
        }
        if (l == 0) return null;
        k = (l - k % l) % l;
        if (k == 0) return head;
        ptr = head;
        var tail2 = ptr;
        while (k != 0) {
            tail2 = ptr;
            ptr = ptr.next;
            k--;
        }
        tail.next = head;
        tail2.next = null;
        return ptr;
    }
}
